File: api/src/Domain/Users/Security/Services/PermissionService.cs
using System;
using System.Collections.Generic;
using System.Threading.Tasks;
using DetailingArsenal.Domain.Users;

namespace DetailingArsenal.Domain.Users.Security {
    public interface IPermissionService : IService {
        Task<List<Permission>> GetAll();
        Task<Permission> GetById(Guid id);
        Task<PermissionSet> GetForRoles(IEnumerable<Role> roles);

        Task<Permission> Create(PermissionCreate create, User user);
        Task Update(Permission permission, PermissionUpdate update);
        Task Delete(Permission permission);
    }

    public class PermissionService : IPermissionService {
        PermissionUniqueSpecification permissionUniqueSpec;
        PermissionNotInUseSpecification permissionNotInUseSpec;
        IPermissionRepo repo;

        public PermissionService(PermissionUniqueSpecification permissionUniqueSpec, PermissionNotInUseSpecification notInUseSpec, IPermissionRepo repo) {
            this.permissionUniqueSpec = permissionUniqueSpec;
            this.permissionNotInUseSpec = notInUseSpec;
            this.repo = repo;
        }

        public async Task<List<Permission>> GetAll() {
            return await repo.FindAll();
        }

        public async Task<Permission> GetById(Guid id) {
            return await repo.FindById(id) ?? throw new EntityNotFoundException();
        }

        public async Task<PermissionSet> GetForRoles(IEnumerable<Role> roles) {
            return await repo.FindForRoles(roles);
        }

        public async Task<Permission> Create(PermissionCreate create, User user) {
            var p = Permission.Create(
                create.Action,
                create.Scope
            );

            await permissionUniqueSpec.CheckAndThrow(p);

            await repo.Add(p);
            return p;
        }

        public async Task Update(Permission permission, PermissionUpdate update) {
            permission.Action = update.Action;
            permission.Scope = update.Scope;

            await permissionUniqueSpec.CheckAndThrow(permission);
            await repo.Update(permission);
        }

        public async Task Delete(Permission permission) {
            await permissionNotInUseSpec.CheckAndThrow(permission);
            await repo.Delete(permission);
        }
    }
}

File: app/src/api/user/security/services/permission-service.ts
import { http } from '@/api/core/http';
import { Permission } from '../data-transfer-objects/permission';
import { PermissionCreate } from '../data-transfer-objects/permission-create';
import { PermissionUpdate } from '../data-transfer-objects/permission-update';

export class PermissionService {
    async getPermissions() {
        const res = await http.get<any[]>('/security/permission');
        return res.data.map(r => this._map(r));
    }

    async createPermission(create: PermissionCreate) {
        const res = await http.post('/security/permission', create);
        return this._map(res.data);
    }

    async updatePermission(update: PermissionUpdate) {
        const res = await http.put(`/security/permission/${update.id}`, update);
        return this._map(res.data);
    }

    async deletePermission(permission: Permission) {
        await http.delete(`/security/permission/${permission.id}`);
    }

    _map(data: any) {
        const p = new Permission(data.id, data.action, data.scope);
        return p;
    }
}
